stdlib/public/core/UnicodeScalarProperties.swift

```diff
@@ -0,0 +1,1449 @@
+) rethrows -> Int {
+  let z = try body(storage)
+  let correctSize = Int(z)
+  if correctSize > storage.capacity {
+    // If the buffer wasn't large enough, replace it with one that is the
+    // correct size and execute the closure again.
+    storage = _SwiftStringStorage<CodeUnit>.create(
+      capacity: correctSize,
+      count: correctSize)
+    return Int(try body(storage))
+  } else {
+    // Otherwise, the buffer has been populated; update its count to the correct
+    // value.
+    storage.count = correctSize
+    return correctSize
+  }
+}
+
+extension Unicode.Scalar.Properties {
+
+  internal func _scalarName(
+    _ choice: __swift_stdlib_UCharNameChoice
+  ) -> String? {
+    let initialCapacity = 256
+
+    var storage = _SwiftStringStorage<UTF8.CodeUnit>.create(
+      capacity: initialCapacity,
+      count: 0)
+    var err = __swift_stdlib_U_ZERO_ERROR
+
+    let correctSize = _expandingStorageIfNeeded(&storage) { storage in
+      return storage.start.withMemoryRebound(
+        to: Int8.self,
+        capacity: storage.capacity
+      ) { storagePtr in
+        err = __swift_stdlib_U_ZERO_ERROR
+        return __swift_stdlib_u_charName(
+          _value, choice, storagePtr, Int32(storage.capacity), &err)
+      }
+    }
+
+    guard err.isSuccess && correctSize > 0 else {
+      return nil
+    }
+    return String(_storage: storage)
+  }
+
+  /// The published name of the scalar.
+  ///
+  /// Some scalars, such as control characters, do not have a value for this
+  /// property in the UCD. For such scalars, this property will be nil.
+  ///
+  /// This property corresponds to the `Name` property in the
+  /// [Unicode Standard](http://www.unicode.org/versions/latest/).
+  public var name: String? {
+    return _scalarName(__swift_stdlib_U_UNICODE_CHAR_NAME)
+  }
+
+  /// The normative formal alias of the scalar, or nil if it has no alias.
+  ///
+  /// The name of a scalar is immutable and never changed in future versions of
+  /// the Unicode Standard. The `nameAlias` property is provided to issue
+  /// corrections if a name was issued erroneously. For example, the `name` of
+  /// U+FE18 is "PRESENTATION FORM FOR VERTICAL RIGHT WHITE LENTICULAR BRAKCET"
+  /// (note that "BRACKET" is misspelled). The `nameAlias` property then
+  /// contains the corrected name.
+  ///
+  /// This property corresponds to the `Name_Alias` property in the
+  /// [Unicode Standard](http://www.unicode.org/versions/latest/).
+  public var nameAlias: String? {
+    return _scalarName(__swift_stdlib_U_CHAR_NAME_ALIAS)
+  }
+}
+
+extension Unicode.Scalar.Properties {
+
+  /// The lowercase mapping of the scalar.
+  ///
+  /// This property is a `String` because some mappings may transform a single
+  /// scalar into multiple scalars. For example, the character "İ" (U+0130
+  /// LATIN CAPITAL LETTER I WITH DOT ABOVE) becomes two scalars (U+0069 LATIN
+  /// SMALL LETTER I, U+0307 COMBINING DOT ABOVE) when converted to lowercase.
```

Comment:
And it's not a `Character` because it can change a single scalar into multiple graphemes.

Reply:
Updated.